### PR TITLE
Activate k8s rbac

### LINKED ISSUE #680: Request Help to Specify --authorization-mode in API Model for k8s RBAC
**Is this a request for help?**:
yes
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
FEATURE REQUEST

k8s 1.6+ supports RBAC/ABAC. However, it requires the API server start with extra parameters, e.g. --authorization-mode=RBAC,ABAC --authorization-policy-file=mypolicy.json

Currently, there is no way to specify them in the API model.

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
Kubernetes 1.6+

**What happened**:
I'm unable to enable the RBAC for my cluster. 

**What you expected to happen**:
In API model, I can define the --authorization-mode or --authorization-policy-file for apiserver.

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:
You can specify multiple mode in the --authorization-mode, e.g. --authorization-mode=RBAC,ABAC. It's important for RBAC, since RBAC requires initial super user to setup the other roles.

### PR DESCRIPTION
**Pull request by Globegitter**
**What this PR does / why we need it**:
Activates RBAC - this seems to be the way forward with k8s.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

Partially fixes #680 as it enables RBAC per default but does not help for upgrading existing clusters.

**Special notes for your reviewer**:
This haven't spent too much time to check that this is the most ideal way to set RBAC up, but it gets it working for now in our test cluster and thought I would share that work.

This depends on #909 which upgrades heapster to v1.4.0

I tested this on Azure on northeurope with a few of our services and seems to be all working.

##
**Release note**:

```release-note
```

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/918)

### CODE REVIEW

File: parts/kubernetesmaster-kube-apiserver.yaml [hunk 1]
```diff
@@ -31,6 +31,7 @@ spec:
         - "--service-account-key-file=/etc/kubernetes/certs/apiserver.key"
         - "--storage-backend=etcd2"
         - "--v=4"
+        - "--authorization-mode=RBAC"
```
Comment (lachie83) [inline parts/kubernetesmaster-kube-apiserver.yaml]: Can you please conditionalize this and bubble it up into the API model? User should explicitly turn RBAC on. I can raise a PR against your branch with the change

File: parts/kubernetesmasteraddons-heapster-deployment.yaml [hunk 2]
```diff
@@ -50,21 +76,20 @@ spec:
         beta.kubernetes.io/os: linux
     metadata:
       labels:
+        task: monitoring
```
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: What's the purpose of this label?

File: parts/kubernetesmasteraddons-heapster-deployment.yaml [hunk 3]
```diff
@@ -24,11 +52,9 @@ spec:
         - "--extra-cpu=0.5m"
         - "--memory=140Mi"
         - "--extra-memory=4Mi"
-        - "--threshold=5"
```
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: Have these three parameters been deprecated?

File: parts/kubernetesmasteraddons-heapster-service.yaml [hunk 4]
```diff
@@ -2,6 +2,7 @@ apiVersion: v1
 kind: Service
 metadata:
   labels:
+    task: monitoring
```
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-service.yaml]: What's the purpose of this label?

File: parts/kubernetesmaster-kube-controller-manager.yaml [hunk 5]
```diff
@@ -23,6 +23,7 @@ spec:
         - "--root-ca-file=/etc/kubernetes/certs/ca.crt"
         - "--service-account-private-key-file=/etc/kubernetes/certs/apiserver.key"
         - "--leader-elect=true"
+        - "--use-service-account-credentials"
```
Comment (lachie83) [inline parts/kubernetesmaster-kube-controller-manager.yaml]: Should this be conditionalized too based on having RBAC enabled?

**Comment by lachie83** [comment]
@Globegitter Thanks for the contribution!

File: pkg/acsengine/pki.go [hunk 6]
```diff
@@ -62,18 +62,22 @@ func CreatePki(extraFQDNs []string, extraIPs []net.IP, clusterDomain string, caP
...
-		apiServerCertificate, apiServerPrivateKey, err = createCertificate("apiserver", caCertificate, caPrivateKey, true, extraFQDNs, extraIPs)
+		apiServerCertificate, apiServerPrivateKey, err = createCertificate("apiserver", caCertificate, caPrivateKey, true, extraFQDNs, extraIPs, nil)
 		errors <- err
 	}()
 
 	go func() {
 		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
 		clientCertificate, clientPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, nil, nil)
```
Comment (lachie83) [inline pkg/acsengine/pki.go]: Looks like you're not passing the organization slice into createCertificate

File: pkg/acsengine/pki.go [hunk 7]
```diff
@@ -62,18 +62,22 @@ func CreatePki(extraFQDNs []string, extraIPs []net.IP, clusterDomain string, caP
...
 		errors <- err
 	}()
 
 	go func() {
 		var err error
+		organization := make([]string, 1)
+		organization[0] = "system:masters"
 		kubeConfigCertificate, kubeConfigPrivateKey, err = createCertificate("client", caCertificate, caPrivateKey, false, nil, nil)
```
Comment (lachie83) [inline pkg/acsengine/pki.go]: Looks like you're not passing the organization slice into createCertificate

**Comment by lachie83** [comment]
@Globegitter I've submitted a PR against your branch to update this PR.

https://github.com/Globegitter/acs-engine/pull/1

**Comment by lachie83** [comment]
The only issue remaining is the kube-addon-manager is a static pod that needs a serviceAccount and associated ClusterRoleBinding to start. We need to think about the best way to bootstrap that.

**Comment by lachie83** [comment]
I've gotten kube-addon-manager to work without using a service account. We should be good to go

File: pkg/acsengine/pki.go [hunk 8]
```diff
@@ -62,18 +62,22 @@ func CreatePki(extraFQDNs []string, extraIPs []net.IP, clusterDomain string, caP
 
 	go func() {
 		var err error
-		apiServerCertificate, apiServerPrivateKey, err = createCertificate("apiserver", caCertificate, caPrivateKey, true, extraFQDNs, extraIPs)
+		apiServerCertificate, apiServerPrivateKey, err = createCertificate("apiserver", caCertificate, caPrivateKey, true, extraFQDNs, extraIPs, nil)
 		errors <- err
```
Comment (seanknox) [inline pkg/acsengine/pki.go]: We need some unit tests on this, too.

**Comment by seanknox** [comment]
Thanks @Globegitter and @lachie83! Can you rebase and we'll try to get this in soon.

**Comment by lachie83** [comment]
@Globegitter pulled another commit to your branch. https://github.com/Globegitter/acs-engine/pull/2

**Comment by lachie83** [comment]
Removed dep on #909. Although now merged

**Comment by lachie83** [comment]
Pushed rebase into that PR as well

**Comment by lachie83** [comment]
I've tested with 1.6.6 with both RBAC on and off and we should be GTG


Here are the tests:

* All log files of pods in kube-system for errors
* HPA working (Heapster)
* Service discovery working (kube-dns) Resolve a service from within a pod
* Dashboard accessible via kube-proxy and resources available

File: parts/kubernetesmasteraddons-heapster-deployment.yaml (hunk 3 above)
Comment (Globegitter) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: I am quite sure - I can try and find the source again where I saw that.

File: parts/kubernetesmasteraddons-heapster-service.yaml (hunk 4 above)
Comment (Globegitter) [inline parts/kubernetesmasteraddons-heapster-service.yaml]: Just when I took things from the heapster repo they had it. Haven't thought too much further into it.

File: pkg/acsengine/pki.go (hunk 8 above)
Comment (Globegitter) [inline pkg/acsengine/pki.go]: Will do asap.

File: parts/kubernetesmaster-kube-addon-manager.yaml [hunk 9]
```diff
@@ -20,6 +19,7 @@ spec:
     - name: msi
       mountPath: "/var/lib/waagent/ManagedIdentity-Settings"
       readOnly: true
+  serviceAccountName: kube-addon-manager
```
Comment (lachie83) [inline parts/kubernetesmaster-kube-addon-manager.yaml]: This needs to be removed.

File: parts/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 10]
```diff
@@ -123,5 +140,10 @@ spec:
             cpu: 10m
             memory: 50Mi
       dnsPolicy: Default
+<<<<<<< HEAD
```
Comment (lachie83) [inline parts/kubernetesmasteraddons-kube-dns-deployment.yaml]: This needs to be updated

File: pkg/acsengine/const.go [hunk 11]
```diff
@@ -84,11 +84,19 @@ var KubeImages = map[api.OrchestratorVersion]map[string]string{
 		"hyperkube":    "hyperkube-amd64:v1.7.0",
 		"dashboard":    "kubernetes-dashboard-amd64:v1.6.1",
 		"exechealthz":  "exechealthz-amd64:1.2",
+<<<<<<< HEAD
```
Comment (lachie83) [inline pkg/acsengine/const.go]: Also this

File: pkg/api/vlabs/types.go [hunk 12]
```diff
@@ -139,6 +139,7 @@ type KubernetesConfig struct {
 	ClusterSubnet                    string  `json:"clusterSubnet,omitempty"`
 	NetworkPolicy                    string  `json:"networkPolicy,omitempty"`
 	DockerBridgeSubnet               string  `json:"DockerBridgeSubnet,omitempty"`
+	EnableRbac          			 bool    `json:"enableRbac,omitempty"`
```
Comment (lachie83) [inline pkg/api/vlabs/types.go]: Remove one of these

File: pkg/api/types.go [hunk 13]
```diff
@@ -134,6 +134,7 @@ type KubernetesConfig struct {
 	ClusterSubnet                    string  `json:"clusterSubnet,omitempty"`
 	NetworkPolicy                    string  `json:"networkPolicy,omitempty"`
 	DockerBridgeSubnet               string  `json:"dockerBridgeSubnet,omitempty"`
+	EnableRbac          			 bool    `json:"enableRbac,omitempty"`
```
Comment (lachie83) [inline pkg/api/types.go]: Remove one of these

File: pkg/api/convertertoapi.go [hunk 14]
```diff
@@ -569,6 +569,7 @@ func convertVLabsKubernetesConfig(vlabs *vlabs.KubernetesConfig, api *Kubernetes
 	api.ClusterSubnet = vlabs.ClusterSubnet
 	api.NetworkPolicy = vlabs.NetworkPolicy
 	api.DockerBridgeSubnet = vlabs.DockerBridgeSubnet
+	api.EnableRbac = vlabs.EnableRbac
```
Comment (lachie83) [inline pkg/api/convertertoapi.go]: Remove one of these

File: pkg/api/converterfromapi.go [hunk 15]
```diff
@@ -548,6 +548,7 @@ func convertKubernetesConfigToVLabs(api *KubernetesConfig, vlabs *vlabs.Kubernet
 	vlabs.ClusterSubnet = api.ClusterSubnet
 	vlabs.NetworkPolicy = api.NetworkPolicy
 	vlabs.DockerBridgeSubnet = api.DockerBridgeSubnet
+	vlabs.EnableRbac = api.EnableRbac
```
Comment (lachie83) [inline pkg/api/converterfromapi.go]: Remove one of these

File: parts/kubernetesmasteraddons-heapster-service.yaml (hunk 4 above)
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-service.yaml]: Okay.

**Comment by lachie83** [comment]
@Globegitter I've also created #1043. Totally no pressure but I'm happy to help get the awesome work you've done across the line. Also happy to work through this PR if you're happy to do that. I'll leave it up to you.

**Comment by Globegitter** [comment]
@lachie83 Sorry just had little time yesterday. I am happy to finish this off - just having trouble with running tests locally to write them for pki.go. It seems to have trouble because everything is namespaced to 'github.com/Azure/' and my fork is namespaced to 'github.com/Globegitter/'. How is that supposed to work locally on a fork?

**Comment by karataliu** [comment]
@Globegitter I think you can try copy your local repo to '$GOPATH/src/github.com/Azure' and continue the deployment, you'll still be able to push to your own fork since git remote config is there.

BTW, I've tried the PR and found some built-in containers such as dns, dashboard are not started by default. While after a master node reboot, everything comes back again. Could you please help re-verify?

**Comment by martell** [comment]
Looking forward to getting this support in
`po/heapster-v1.2.0-2517007820-6jncm                1/2       CrashLoopBackOff    7          15m`
I had updated to 1.4.0 locally but because this PR was already created I left it.

File: pkg/acsengine/pki.go (hunk 8 above)
Comment (Globegitter) [inline pkg/acsengine/pki.go]: There where no unit tests for pki.go so added a new pki_test but just testing the changes that I introduced.

**Comment by lachie83** [comment]
> Looking forward to getting this support in
po/heapster-v1.2.0-2517007820-6jncm 1/2 CrashLoopBackOff 7 15m
I had updated to 1.4.0 locally but because this PR was already created I left it.

Why is 1.2 being deployed off the changes in this PR or are you referring to another issue?

**Comment by lachie83** [comment]
I will retest this today then we should be gtg

**Comment by lachie83** [comment]
@karataliu 

>BTW, I've tried the PR and found some built-in containers such as dns, dashboard are not started by default. While after a master node reboot, everything comes back again. Could you please help re-verify?

I have seen this before as well and looked into the issue. The controller-manager stoppped functioning. I wonder if there is some sort of race condition in `--use-service-account-credentials`

File: parts/kubernetesmasteraddons-kube-dns-deployment.yaml [hunk 16]
```diff
@@ -8,15 +18,19 @@ metadata:
   name: kube-dns-v20
   namespace: kube-system
 spec:
+  strategy:
```
Comment (lachie83) [inline parts/kubernetesmasteraddons-kube-dns-deployment.yaml]: Please remove this and split into a another PR if you think it's needed

File: parts/kubernetesmasteraddons-heapster-service.yaml (hunk 4 above)
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-service.yaml]: Please remove this. And break out into another PR

File: parts/kubernetesmasteraddons-heapster-deployment.yaml (hunk 3 above)
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: I found it. They have been removed if you go to v2.0 which we are. All good.

File: pkg/acsengine/pki.go (hunk 8 above)
Comment (lachie83) [inline pkg/acsengine/pki.go]: Perfect. That's fine!

File: pkg/acsengine/pki_test.go [hunk 17]
```diff
@@ -0,0 +1,77 @@
+package acsengine
```
Comment (lachie83) [inline pkg/acsengine/pki_test.go]: Tests look good. Thanks for adding!

**Comment by martell** [comment]
> Why is 1.2 being deployed off the changes in this PR or are you referring to another issue?
When I was trying 1.7.1 at the weekend I noticed that because there is no config for 1.4.0 it defaults back to 1.2.0. I believe this is because of the `--deployment` flag in `parts/kubernetesmasteraddons-heapster-deployment.yaml`
This PR fixes this.
Note: 1.6.7 fixes cross federation issues with rbac which could be very useful for end users along with addressing a bunch of CVE's

**Comment by lachie83** [comment]
@Globegitter Only issue I have is with the heapster-nanny. Raised this to fix. https://github.com/Globegitter/acs-engine/pull/3

**Comment by lachie83** [comment]
Just two small nit comments left above and then we are LGTM.

I've tested both 1.6.x and 1.7.x with RBAC enabled and disabled.

**Comment by Globegitter** [comment]
@lachie83 Should be all done now.

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by Globegitter** [comment]
@lachie83 Anything faling here? Can't see the jenkins results.

**Comment by martell** [comment]
Hey @brendandburns :)
Could you add @Globegitter and myself to the read permissions of the public jenkins bot?
Currently getting `martell is missing the Overall/Read permission`

**Comment by martell** [comment]
@Globegitter: can you rebase on master (maybe squash some if possible)? I was doing a rebase for local testing but there are a bunch of conflicts and 3 way merges. I believe it is because there is a PR in a PR. I managed to get through it but would prefer if you (the author) done it so I didn't screw anything up. Looking to try this out in combination with 1.7.2

**Comment by lachie83** [comment]
@martell @Globegitter I iwill work on getting you CI access sorted out. I've just taken a look at the build logs and I don't think the issue is related to this set of changes but rather some infrastructure flakiness. I'm also working to get this merged in

**Comment by lachie83** [comment]
@dmitsh Are you able to take a look at why @martell cannot see the build logs?

**Comment by Globegitter** [comment]
@martell Hmm, I just merged master and it applied without any issues.

**Comment by martell** [comment]
`git rebase origin/master` shows the issues.
did you update your upstream?
it seems this PR is editing files already deleted etc.
```
➜  acs-engine git:(k8s-rbac) git rebase master
First, rewinding head to replay your work on top of it...
Applying: Added minimal rbac setup for k8s.
Using index info to reconstruct a base tree...
M       parts/kubernetesmaster-kube-apiserver.yaml
M       pkg/acsengine/defaults.go
M       pkg/acsengine/pki.go
A       pkg/acsengine/templates.go
Falling back to patching base and 3-way merge...
CONFLICT (modify/delete): pkg/acsengine/templates.go deleted in HEAD and modified in Added minimal rbac setup for k8s.. Version Added minimal rbac setup for k8s. of pkg/acsengine/templates.go left in tree.
Auto-merging pkg/acsengine/pki.go
Auto-merging pkg/acsengine/defaults.go
Auto-merging parts/kubernetesmaster-kube-apiserver.yaml
error: Failed to merge in the changes.
Patch failed at 0001 Added minimal rbac setup for k8s.
The copy of the patch that failed is found in: .git/rebase-apply/patch
When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".
```

**Comment by Globegitter** [comment]
@martell ahh you are having a conflict with templates.go - that has been added to .gitignore recently and removed from the repo: https://github.com/Azure/acs-engine/pull/1045 So that should not be giving any conflicts anymore. Maybe manually remove it from your local repo?

File: parts/kubernetesmasteraddons-heapster-deployment.yaml [hunk 18]
```diff
@@ -1,34 +1,103 @@
...
+      - key: "CriticalAddonsOnly"
+        operator: "Exists"
+      serviceAccountName: heapster
       containers:
       - image: <kubernetesHeapsterSpec>
         command:
         - "/heapster"
         - "--source=kubernetes.summary_api:\"\""
         name: heapster
-        resources:
```
Comment (brendandburns) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: Please don't remove the resource limit section.

File: parts/kubernetesmasteraddons-heapster-deployment.yaml [hunk 19]
```diff
@@ -1,34 +1,103 @@
...
   template:
+    metadata:
+      labels:
+        k8s-app: heapster
+        version: v1.4.0
+      annotations:
+        scheduler.alpha.kubernetes.io/critical-pod: ""
     spec:
+      tolerations:
+      - key: "CriticalAddonsOnly"
```
Comment (brendandburns) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: let's drop quotes `"` where we can for readability.

File: parts/kubernetesmasteraddons-heapster-deployment.yaml (hunk 18 above)
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: I removed this as it's taken care of by the pod-resizer but thinking about it more and given some other issues we currently have with the resizer I agree. We should keep the resource limits. I'll put this back

File: parts/kubernetesmasteraddons-heapster-deployment.yaml (hunk 19 above)
Comment (lachie83) [inline parts/kubernetesmasteraddons-heapster-deployment.yaml]: Agreed. I'll update

**Comment by lachie83** [comment]
@Globegitter I have a few clean up items here. I will rebase and submit a PR agains't your branch. Hoping to get this merged in

**Comment by lachie83** [comment]
@Globegitter -- https://github.com/Globegitter/acs-engine/pull/6/

**Comment by lachie83** [comment]
@seanknox @Globegitter once we merge the PR above I think we should be GTG. I have successfully tested against 1.5/6/7 but let's review the changes once that PR is merged in

**Comment by lachie83** [comment]
@Globegitter not sure what happened with that merge. All those commits should be in master

**Comment by lachie83** [comment]
From my branch which should be in-sync with your branch I see -- https://github.com/Azure/acs-engine/compare/master...lachie83:k8s-rbac-updates?expand=1

**Comment by Globegitter** [comment]
@lachie83 I just did a squash and merge, sorry I am just used to that from our dev workflow. I am happy to revert and merge it in normally if that is how I should do it for this repo. Then I am also happy to resolve the merge conflicts with master.

**Comment by lachie83** [comment]
@Globegitter can you revert that squashed commit and remerge the PR without the squash.

I'm trying to get this PR reviewed today and it would be great if I could get it in the right shape.

Would you have any issues with me reopening this PR with your commits and pushing it across the line?

**Comment by Globegitter** [comment]
@lachie83 happy for you to do this and close this :)

**Comment by seanknox** [comment]
@martell huge thanks for your work on this! 

giphy 5

**Comment by lachie83** [comment]
Big thanks to @Globegitter for this work. Continuing in #1143
